### ISSUE TITLE
[Proposal] Modding/TC system

### ISSUE THREAD

**Issue by Mutos** [naev/naev#1116]
I want to propose here a **modding/TC system**, that had previously been discussed in the forum years ago, but never came through.

First, context... I use NAEV as engine for the SoC game, which thus is technically a **Total Conversion** (TC). It has led me to experience a number of issues, related to the fact that the _**dat**_ folder is monolithic, mixing NAEV-as-engine files and universe-dependant files. As such, I'm always wondering what had changed in NAEV's **_dat_** common files and must be copied to SoC's own **_dat_**, and what I can modify or not while developing SoC's own **_dat_**.

So the goal is to enable TCs and mods to be more easily developed and maintained.
- Having a consistent mod system could help development, as one could easily develop new features in a mod, enable it for test and integrate the features in main dataset when ready;
- Having a TC system could prompt other people wanting to make a NAEV-like game in their own universe to use NAEV as engine just as I do for SoC, which would promote NAEV as an easy-to-use engine and expand the NAEV community.

The idea would be to split the **_dat_** folder into :
- An **_engine_** folder, holding all files "NAEV-as-engine" files, common to all TCs and mods;
- a **_universes_** folder, where each TC developer would create their own subfolder for the TC;
- a **_mods_** folder, where each mod would reside in a subfolder;
- a **_mods.xml_** file would be added to **_dat_** root, to know which universe (only one) and mods (ordered list) are currently active.

Each data folder ("base folder" + "TC's universe subfolder" + "mod subfolder") would have the same structure as the current **_dat_**. When a data file is to be read, it would be first sought for in the base folder, then in the active universe folder, then in the active mods folders in the order listed in **_mods.xml_**, with the search returning the first matching file.

Implementation could come in 3 steps :
- Only implement the search system, the 3 main folders (**_engine_**, **_universes_**, **_mods_**), and the **_mods.xml_** file, leaving everything currently in **_dat_** stored in the **_universes/naev_** folder;
- progressively move files to the **_engine_** folder as they are identified as common. Keep in mind that a TC developer could always override a common file, but would do it willingly and knowingly, opposed to the current "fuzzy" situation;
- once the new data schema is up and running, integrate a mods selector, that would control **_mods.xml_** consistency and allow to change it graphically. It could either be made as an independent launcher, a startup optional first step, an extra main menu item or an extra options menu tab, the latter two solutions requiring NAEV to be restarted for changes to take effect.

What do you think about that ? Particularly, would the implementation of the search system be difficult to do or not ? That comes down to : is there a single function to call when you want a data file from **_dat_**, or does each developer do as they feel in different c files ? The idea would be to make a function to convert a regular dat-path into an actual "first found file" path, and use this function everywhere.

**Comment by Mutos** [comment]
Made a `modding` branch in my `Mutos/naev` forked repo.

However, it will take a long time as I can only work on it little by little. Today is an exception, I'm on leave, at home and have some free time ^-^

**Comment by Mutos** [comment]
Hi all, some news...

Now the `modding` branch still doesn't work, but it goes farther than ever ^-^

For now I'm implementing gross redirection from `dat/` to `dat/universes/naev-vanilla/` and even testing a little of what could be done adding `dat/engine/`, but for now just on one function...

I've chosen to go through `ndata.[ch]` and `nfiles.[ch]` and examining what to do with each function, in the order of appearance while loading NAEV to be able to progressively test everything.

It's a slow process because I've got to understand how all this is laid out, there seems to be many ways to find a file and all of them have to be covered. Would per chance anybody have a schema or text explanation, because for now it's painfully slow. Thanks a lot for any hint ^-^

I'll keep you informed, read you soon !

**Comment by Mutos** [comment]
Hi all,

First seemingly working version of my `modding` branch !!!

I say "seemingly', because :
- It does load and run correctly but I've not tested it extensively;
- When listing a directory, it still returns duplicates when files exist in both Engine and Universe paths.

This is the next fix to implement.

Once this will be fixed and I've checked more extensively how it runs, I'll be able to tackle the issue of how to store current Universe... Rather than adding an XML file, I'd add a line `universe=naev-vanilla` to `conf.lua`, seems to be its rightful place for such a setting...

I'll be glad if someone would give that version a whirl, to check if my code is not too buggy and if it does not break some obscure feature using a function I overlooked ^-^

**Comment by Mutos** [comment]
Feature broken… The map editor doesn't work anymore, it crashes, apparently on some memory issue… My worst nightmare coming true...

**Comment by Mutos** [comment]
Hi all,

Some news from the modding development : 
- No actual development of the source itself;
- Just tried some ways to merge naev/naev master into modding development.

I found that git was smart enough to find moved files and properly update them, so updated `dat/` files end up in the proper folder, be it `dat/engine/` of `dat/universes/naev-vanilla/`. That means a `git merge --no-ff --no-commit` is enough for me, after that I check that dat files were properly moved, then I manually solve any conflict that might exist.

**Comment by bobbens** [comment]
Sorry for the late reply. I'm not sure to what extend this is all necessary. Let me give my 2 cents:

* First of all, different games can be done using the same binary by using "-d" to specify the data path. This would let you do something like `./naev -d default-ndata.zip` and `./naev -d tc-ndata.zip`. It does overwrite the config, but it should also be easily possible to add a no update path flag to not update the path in the config file.
* Regarding mods, this is something that has been asked for a lot, the official stance has been that we would rather have new content in the mainline data rather than have separate mods for everything. That said, it will probably be implemented down the line, the issue is that doing it right is a lot of work, and you'll inevitable get into lots of conflicts and issues. The best step forward would be to get rid of monolithic files (e.g., `dat/mission.xml`, `dat/commodity.xml`, `dat/event.xml`, ...). Then, it should be straightforward to basically concatenate all the core data and mod data and loop over it.

When developing a TC, I think the easiest way would be to:
1. Clone naev
2. Compile
3. Create new repository and put it in `naev/my-tc`
4. Have the repository only consist of data files.
5. Run naev with `../naev` from `naev/my-tc` to run with the new data.

If there are code changes, the above would be the same, except that the naev wouldn't be the original repository, but a modified fork. By not actually deleting the `dat/` you would never have problems with merging stuff in `dat/`.

**Comment by Mutos** [comment]
Hello bobbens,

Thanks for your reply. OK for mods, those were extra and maybe dangerous in my view as in yours ^-^

As for my development of SoC, merging `dat/` proved to be a true nightmare and several times, I was stuck and needed to completely start anew, roughly as you describe, then painfully merge all SoC changes back into original `dat/`. So for me personally and for SoC, the `dat/universes/` feature is not an option. I just won't be able to keep working if it's not implemented. Otherwise I wouldn't have devoted so much time for that ^-^

The other idea is that it will also enable to built a TC more easily. Your process is right, but finding what to change for building a TC isn't easy, to tell it blankly. Everything is so interlinked, that 

The third benefit will be to have a better coding, I would decribe as "layered"... The idea is quite classic, it's all about setting true engine files apart from universe-related files and splitting files mixing both parts. The split would typically result in an engine file making the hard, neutral work, and a universe header setting all universe-specific values. I'm thinking mainly about Lua file, for instance some mission and event scripts should be faction-neutral but somehow they do have Empire or Proteron references in them...

**Comment by bobbens** [comment]
OK, as an exercise, I've created a "minimal" ndata to actually launch naev and start a new game. The repository is at https://github.com/naev/minimal-ndata . It's not as bad as I thought, but there is quite a bit of stuff (shaders, etc.) that shouldn't be part of the ndata and should probably be embedded in the binary. Afterwards, if we get rid of the remaining monolithic files and some of the not good hardcoded behaviour (move to `dat/start.xml`? or rename it `dat/ndata.xml`?) it shouldn't be too bad. When it is streamline it can be put into `extras/` in the main repo and then you can do what I mentioned above, which is create the ndata as a separate repo, and just launch naev with `./naev my_new_ndata.zip`. It should then be fairly trivial to maintain, just don't keep an entire copy of naev. If you want to edit the C code, keep a naev fork, but maintain the data in a separate repo to avoid merge issues.

To summarize what has to be improved is:
1. Move shaders into binary.
2. Get rid of monolithic files (dat/*.xml).
3. Move more hardcoded paths and files into `dat/start.xml` that should get renamed `dat/ndata.xml`.
4. Add minimal example to `extras/

**Comment by Mutos** [comment]
Hi bobbens,

Yes, I understand what you say.

The issue for me is, for instance, files like the AI or spawn systems. They evolve with time in NAEV and sometimes they're linked with C code, so I must keep them up-to-date on SoC if I want SoC to stay compatible with NAEV.

For instance, today, SoC is not compatible anymore with NAEV, I'm stuck with an old NAEV for SoC and I just can't get the bug, it's somewhere in either the AI or spawn system, but I just can't nail it. It's like having a permanent Damocles sword upon the SoC project, because if I can't quickly get back in the train, SoC will become hopelessly obsolete.

**Comment by Mutos** [comment]
Hi bobbens, I tried to get https://github.com/naev/minimal-ndata, but it seems not to exist.

**Comment by bobbens** [comment]
> Hi bobbens, I tried to get https://github.com/naev/minimal-ndata, but it seems not to exist.

Sorry, had it set to private.

> The issue for me is, for instance, files like the AI or spawn systems. They evolve with time in NAEV and sometimes they're linked with C code, so I must keep them up-to-date on SoC if I want SoC to stay compatible with NAEV.

This is somewhat inevitable, in the sense that the spawn scripts and AI scripts should probably be changed depending on the data. AI can be more or less used as is, however, spawning stuff does pose more of an issue.

> That's why my idea was to split NAEV data files between engine and universe, keep a single engine and choose the universe you want to play. It's not about creating a TC, it's all about creating a maintainable TC able to follow NAEV in its development.

Compared with other games I would argue we are more "modular". That said, fixing the issues I mentioned above should make it much more compatible. That said, until version 1.0.0, I don't think we can guarantee that there won't be any future breakage.

**Comment by Mutos** [comment]
Hi bobbens,

First thanks for the time you spent on this, as I'm the only one concerned here by this issue, except if there's other TCers unknown to me ^-^

I'll try the `-d` flag to begin with, this should simplify my launching scripts.

As for the concatenation of data files, I guess you mean having several smaller files in a folder and read them in a loop, instead of just reading one file ? For instance, replacing `faction.xml` by a `factions/definitions` folder with for instance a `generic.xml` file containing `Trader`, `Civilian`, `Independent` and `Pirate` factions, and a `naev.xml` containing `Empire`, `Proteron`, `Thurion`, `Za'lek` and so on, and maybe a `mission-specific.xml` holding `Nasin` and `Thugs`... and for SoC I would replace `naev.xml` by a `soc.xml` containing `Core League`, `Hoshi no Hekka`, `Shoden Clusters` and so on, or maybe having several smaller files for `human.xml`, `sshaad.xml`, `rith.xml` and so on...

**Comment by bobbens** [comment]
> I'll try the -d flag to begin with, this should simplify my launching scripts.

Sorry, it was just `./naev my_ndata.zip`. You don't need `-d`

> As for the concatenation of data files, I guess you mean having several smaller files in a folder and read them in a loop, instead of just reading one file ? For instance, replacing faction.xml by a factions/definitions folder with for instance a generic.xml file containing Trader, Civilian, Independent and Pirate factions, and a naev.xml containing Empire, Proteron, Thurion, Za'lek and so on, and maybe a mission-specific.xml holding Nasin and Thugs... and for SoC I would replace naev.xml by a soc.xml containing Core League, Hoshi no Hekka, Shoden Clusters and so on, or maybe having several smaller files for human.xml, sshaad.xml, rith.xml and so on...

Each faction or whatever would have it's own file like we do with `ssys/`, `ships/`, and `outfits/`. Each folder would be unique to each ndata. This should basically remove any issues with merging forever (unless everyone edits same files), and should be much for flexible if some day mods are done.

> As for AI and spawning, I generally try to use them as is, just copying some NAEV faction scripts into a new SoC faction script and adapting them to suit SoC's particular context. The issue with that is, once the file is forked, there's no easy way to cascade changes done to the original file into all forked ones. That's why I think universe files should avoid to contain generic logic. Easier said than done, though ^-^

I think there is no way to avoid this, other than to try to use the Naev common files as much as possible, and update it when they are updated in Naev.

**Comment by Mutos** [comment]
OK, got it for the false flag and multiple XML files ^-^ thanks for the explanation !

As for AI and other things like that, the common files are the key, that's why I initially thought to group them in an engine folder...

**Comment by bobbens** [comment]
> As for AI and other things like that, the common files are the key, that's why I initially thought to group them in an engine folder...

I want to keep it to binary + single zip. So I guess the best way to "handle" this would be to include the common code as a sort of library (`dat/ai/include/*`) that you could use as a base for the AI in the module. This could be done with other important parts of the code. They would be embedded in the binary as strings and be auto-included.

**Comment by UncombedCoconut** [comment]
I'm just noticing this discussion. (I joined somewhat later.) How does all of this look in light of PhysicsFS and its separate issue trackers? Should some of these issues be combined?
There are definitely still monolithic XML files to split (that's https://github.com/naev/naev/issues/1646).
It's probably time to just stop saving data paths into the user's config file. If someone's existing setup depends on that, we probably broke it post-0.8 anyway. :)
There are still Lua refactorings (splitting core functionality from factions'/missions' usage) to explore. I'd expect this change to be gradual.

A mod could definitely ship as one zip file that gets used as a data overlay.

**Comment by Mutos** [comment]
Hi all,


Thansk @UncombedCoconut for jumping in the discussion and proposing useful ideas ^-^

I've seen and tested the new PhysFS system, as you may have guessed by the issues and questions I've been asking lately, it works nicely and goes a great step towards the goal !

Now how about splitting data files into engine and universe ? @bobbens, I understand your binary + ndata principle, but couldn't it be just extended to binary + engine data + universe data ? Then any TC would just replace the universe data file by its own.

If you're OK with that principle, the main remaining step would be getting rid of the monolithic XML files, that is, what we were talking about a few posts up, replacing them with a folder and a file per object. I scanned issues for this, but it doesn't seem to exist. Is it OK to create and issue for that to track this action ?

**Comment by UncombedCoconut** [comment]
Hi, first I'll reply to the easy part:

> If you're OK with that principle, two main remaining steps would be getting rid of :
> 
>     * Monolithic XML files : what we were talking about a few posts up, replacing them with a folder and a file per object;
> 
>     * Mixed engine-universe data files : as @UncombedCoconut said, about Lua refactoring, I see that mostly for AI, missions/events, and spawn, and on many of these areas it's already quite under way if I recall well.
> 
> 
> I scanned issues for this, but it doesn't seem to exist. Is it OK to create issues to track these two actions ?

Yes. So far these ideas are buried in comments/discussions. XML splitting especially makes a good stand-alone tracker; there's no question about what the steps are and what our progress is.
Engine/universe split is perhaps murkier / to be informed by design discussions...

**Comment by bobbens** [comment]
> Now how about splitting data files into engine and universe ? @bobbens, I understand your binary + ndata-ZIP principle, couldn't it be just extended to binary + engine-data-ZIP + universe-data-ZIP ? Then any TC would just replace the universe data file by its own and any (albeit optional and further down the line) mod could just add an extra mod-data-ZIP file.

As mentioned by @UncombedCoconut , it's a bit murkier what is what. In particular, what would engine-data consist of? I understand you're thinking about mainly AI, but there's also GLSL (some is required), and a ton of libraries in `dat/script/` these days. For AI, I can imagine most TC would want to use as much as possible, but there's a lot of "Naev" stuff hardcoded in there. Would need some in-depth thoughts to see what is necessary and what isn't.

>     * Monolithic XML files : what we were talking about a few posts up, replacing them with a folder and a file per object;

This is something that has to be done just to avoid conflicts and internal issues. We're almost there too.

> >     * Mixed engine-universe data files : as @UncombedCoconut said, about Lua refactoring, I see that mostly for AI, missions/events, and spawn, and on many of these areas it's already quite under way if I recall well.

As mentioned, this is very tricky because there is no clear distinction between engine and non-engine for lots of items. For example, should all the Love2D API and visual novel framework be treated as what? It does seem like it'll probably end up being some sort of white/blacklisting approach, but even then, it will be messy to be fully cleared. Might be just easier to include as much as possible and let the TC or whatever just overwrite stuff as they want.

**Comment by UncombedCoconut** [comment]
> Hi all,
> 
> Thansk @UncombedCoconut for jumping in the discussion and proposing useful ideas ^-^
> 
> I've seen and tested the new PhysFS system, as you may have guessed by the issues and questions I've been asking lately, it works nicely and goes a great step towards the goal !
> 
> Now how about splitting data files into engine and universe ? @bobbens, I understand your binary + ndata-ZIP principle, couldn't it be just extended to binary + engine-data-ZIP + universe-data-ZIP ? Then any TC would just replace the universe data file by its own and any (albeit optional and further down the line) mod could just add an extra mod-data-ZIP file.

Hmm. PhysicsFS gives us a single logical file system and unified access method. Let's build on that.
The distinction between a zip and a folder is abstracted away. Either one is just a component of the search-path.
Ideally, I'd like mods to be very simple from an engine perspective: a mod is also just a search-path component (= virtual file-system overlay), and a TC is nothing but a mod whose content happens to "totally convert" the game.
In this picture, a mod can't know or care how the virtual file-system (VFS) beneath it is organized on disk, but its author has to worry about the game engine loading incompatible content from the underlying VFS.

UX consideration: I'd rather have installing a mod (e.g. "Voltron Fangame TC") be one reversible step. I don't want to have to turn off a "Sea of Darkness" layer myself.
 to avoid a mashup with the Voltron universe.
Dev consideration: I'd rather not try to draw sharp boundaries between core and universe files, particularly while Naev is seeing heavy development. A creative enough mod may blur the lines anyway. (Formations are a core engine feature. However, the Voltron Fangame TC extends it so fleets can assemble into giant robots.)
Modder consideration: One way or another, if the user installs my TC, I can't have the "Naev: Sea of Darkness" star systems loading into my universe. There's no hiding the issue of suppressing content: I may outright want Dvaered/Sirius kept off the player's standings screen because they aren't of my universe.

What I'm thinking at this point is that we need a system where mods have physfs semantics (so, mainly additive but able to shadow individual files), and the engine decides what to load based on VFS contents.
It picks up `start.xml` directly, and unlike today it gets instructed to load all star systems in `sea_of_darkness/ssys` -- or not, if shadowed -- and mods can extend/override the very choice of which paths get loaded.
Even if this proves to be a good idea, there are still more questions to answer, like how monolithic the "start" data should be.
IMO the most important thing we can do is build a system that lets us change our mind about the trickier questions.
It should be compatible with a developing game whose structure is in flux. Where this conflicts with the stability properties a modder would want, it wins anyway, for now.

**Comment by Mutos** [comment]
Hi all,

OK to begin with XML splitting ^-^That's be one less roadblock!

As for engine/universe splitting, for me, in first approach, an engine file would be any file that has no universe dependencies, that is, which does not refer to anything universe-specific. In a second step, there would be a more detailed approach to split files in a generic part and a universe-dependant part.

> As mentioned, this is very tricky because there is no clear distinction between engine and non-engine for lots of items.

I could find examples in missions, events or AI for instance. As for 'Love2D API and visual novel framework', I just don't know anything of them, so I can't have any opinion ^-^

Suppressing content is one reason of the distinction between "mod" and "universe". For now, I won't talk about mods, but only about TCs. That's why only one folder in the `dat/universe` folder is to be used at any one time, as universes are mutually exclusive. If we decide to integrate that into the NAEV UI, there could be a universe selector, based on what subfolders are present under that `dat/universe` folder, stored in the `conf.lua` and requiring NAEV to be restarted for taking effect. Or there could only be a script for launching NAEV on a given universe...

**Comment by Mutos** [comment]
Opened Issue #1707 for "Split monolithic XML files"...

**Comment by UncombedCoconut** [comment]
> As for engine/universe splitting, for me, in first approach, an engine file would be any file that has no universe dependencies, that is, which does not refer to anything universe-specific. In a second step, there would be a more detailed approach to split files in a generic part and a universe-dependant part.

I like this definition / guiding principle.

> Suppressing content is one reason of the distinction between "mod" and "universe". For now, I won't talk about mods, but only about TCs. That's why only one folder in the `dat/universe` folder is to be used at any one time, as universes are mutually exclusive. If we decide to integrate that into the NAEV UI, there could be a universe selector, based on what subfolders are present under that `dat/universe` folder, stored in the `conf.lua` and requiring NAEV to be restarted for taking effect. Or there could only be a script for launching NAEV on a given universe...

If we want to succeed, we should play to the strengths of physfs.
Specifically, it has no concept of hiding/shadowing/exclusion (other than file-level search precedence), nor would that fit its design.
I think we can harmonize your approach with my comments: instead of having an exclusive dat/universe, use namespacing.
The core game can put "universe" content in `dat/sea_of_darkness`.
A TC can put parallel "universe" content in `dat/voltron_fangame`, and override one file (`start.xml`?) which defines the "universe" location. If the user installs this atop Naev (and its universe), the `dat/sea_of_darkness` files will all still exist, but the engine won't load them (unless so directed by a script under the TC author's control, anyway).
As today we'd probably tie this to saved-game-related functionality -- Naev would understand the universes were different.

**Comment by bobbens** [comment]
Small update. Naev now [supports simple plugins](https://github.com/naev/naev/blob/main/docs/manual/sec/intro.md#plugins). However, there is no explicit support for TC yet, you would have to create empty files to mask all engine files, which is not practical. I'm leaning to allowing a simple blacklist folder approach and have the nfile / ndata abstractions use that on top of PHYSFS, and we could add a "TC" option that would have good defaults to hide most unnecessary things. This would have to be defined in a `mod.xml` or something, which we haven't done yet.

**Comment by bobbens** [comment]
Setting this to 0.10.0, with the intention to allow blacklisting files (and TC option). Idea is to use [PHYSFS_registerArchiver](https://icculus.org/physfs/docs/html/physfs_8h.html#ace0ecda857dc622e2d005af02972237d) to create empty files on files that match some sort of regex.

**Comment by bobbens** [comment]
Added support for blacklisting files and explicit support for making a mod a total conversion as described in [the manual](https://github.com/naev/naev/blob/main/docs/manual/sec/plugins.md). This means it should be possible to create simple somewhat future-proof total conversion plugins. However, it's still likely to be rough around the edges. I guess I'll leave this issue open until a minimal total conversion plugin is created as an example.

**Comment by bobbens** [comment]
I've added a simple TC to https://github.com/naev/total_conversion_plugin_example . If you clone it into your plugin directory (~/.local/share/naev/plugins on linux), it should overwrite everything. It has some rough edges (Local System Map not available causes some warnings when landed), but it's a good proof of concept and much better than the minimal ndata repo as all the necessary shaders and extra libraries (including AI) are provided. While there's still a lot of things to do, I think we can close this issue and open smaller issues focusing on things that still have to be improved. Thanks!
